caput/profile.py

```diff
@@ -163,3 +168,156 @@ def __exit__(self, *args, **kwargs):
+        if not self.use_profiler:
+            return
+        self.stop(self._name)
+
+    def start(self, name):
+        """
+        Start profiling.
+        Results generated when `stop` is called are based on this start time.
+        Attributes
+        ----------
+        name : str
+            Description of what is profiled. You have to pass the same str to `stop`.
+        """
+        self._start_time[name] = time.time()
+
+        # Get all stats at the same time
+        with self.oneshot():
+            self._start_cpu_times[name] = self.cpu_times()
+            self.cpu_percent()
+            self._start_memory[name] = self.memory_full_info().uss
+            if psutil.MACOS:
+                self._start_memory[name] = psutil.disk_io_counters()
+            else:
+                self._start_disk_io[name] = self.io_counters()
+
+    def stop(self, name):
+        """
+        Stop profiler and return results.
+        `start` must be called first.
+        Attributes
+        ----------
+        name : str
+            Description of what is profiled. Has to have been passed to `start` before.
+        Returns
+        -------
+        cpu_times : `collections.namedtuple`
+            Same as `psutil.cpu_times`. Process CPU times since `start` was called in seconds.
+        cpu_percent :  float
+            Process CPU utilization since `start` was called as percentage. Can be >100 if multiple threads run on
+            different cores. See `PSUtil.cpu_count` for available cores.
+        disk_io : `collections.namedtuple`
+            Same as `psutil.io_counters` (on Linux) or `psutil.disk_io_counters` (on MacOS).
+            Difference since `start` was called.
+        memory : str
+            Difference of memory in use by this process since `start` was called. If negative,
+            less memory is in use now.
+        Raises
+        ------
+        RuntimeError
+            If stop was called before start.
+        """
+        stop_time = time.time()
+
+        # Get all stats at the same time
+        with self.oneshot():
+            cpu_times = self.cpu_times()
+            cpu_percent = self.cpu_percent()
+            memory = self.memory_full_info().uss
+            if psutil.MACOS:
+                disk_io = psutil.disk_io_counters()
+            else:
+                disk_io = self.io_counters()
+
+        if name not in self._start_cpu_times:
+            raise RuntimeError(
+                f"PSUtilProfiler.stop was called before start for '{name}'."
+            )
+
+        # Construct results
+        CPU_Times = collections.namedtuple("CPUtimes", list(cpu_times._fields))
+        cpu_times = CPU_Times(*np.subtract(cpu_times, self._start_cpu_times.pop(name)))
+        DiskIO = collections.namedtuple("DiskIO", list(disk_io._fields))
+        disk_io = DiskIO(*np.subtract(disk_io, self._start_disk_io.pop(name)))
+        memory = memory - self._start_memory.pop(name)
```

Comment:
> Also, another thought. You wrap the results in a namedtuple, I wonder if it would be better to do it with a dict. I find it useful to iterate over the items without knowing what they are, and you can't do that with a namedtuple without using semi-private attributes (either ._as_dict() or ._fields)